File: app/src/main/java/com/example/injection/module/ServicesModule.java
package com.example.injection.module;

import com.squareup.moshi.KotlinJsonAdapterFactory;
import com.squareup.moshi.Moshi;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import example.com.stackrx.services.questions.service.StackExchangeService;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;

@Module
public class ServicesModule {

    @Singleton
    @Provides OkHttpClient provideOkHttpClient() {
        return new OkHttpClient.Builder().addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BASIC)).build();
    }

    @Singleton
    @Provides
    Moshi provideMoshi() {
        return new Moshi.Builder().add(new KotlinJsonAdapterFactory()).build();
    }

    @Singleton
    @Provides
    StackExchangeService provideStackExchangeService(OkHttpClient client, Moshi moshi) {
        return new StackExchangeService(client, moshi);
    }

}

File: app/src/main/java/com/example/injection/module/ServicesModule.kt
package com.example.injection.module

import com.squareup.moshi.KotlinJsonAdapterFactory
import com.squareup.moshi.Moshi
import dagger.Module
import dagger.Provides
import example.com.stackrx.services.questions.service.StackExchangeService
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import javax.inject.Singleton

@Module
class ServicesModule {

    @Singleton
    @Provides
    internal fun provideOkHttpClient(): OkHttpClient =
            OkHttpClient.Builder().addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BASIC)).build()

    @Singleton
    @Provides
    internal fun provideMoshi(): Moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()

    @Singleton
    @Provides
    internal fun provideStackExchangeService(client: OkHttpClient, moshi: Moshi): StackExchangeService =
            StackExchangeService(client, moshi)

}
